presto-main/src/main/java/com/facebook/presto/resourcemanager/RatisServer.java

```diff
@@ -0,0 +1,104 @@
+ */
+package com.facebook.presto.resourcemanager;
+
+import com.facebook.presto.metadata.InternalNode;
+import com.facebook.presto.metadata.InternalNodeManager;
+import org.apache.ratis.conf.RaftProperties;
+import org.apache.ratis.grpc.GrpcConfigKeys;
+import org.apache.ratis.protocol.RaftGroup;
+import org.apache.ratis.protocol.RaftGroupId;
+import org.apache.ratis.protocol.RaftPeer;
+import org.apache.ratis.protocol.RaftPeerId;
+import org.apache.ratis.server.RaftServer;
+import org.apache.ratis.server.RaftServerConfigKeys;
+import org.apache.ratis.statemachine.impl.BaseStateMachine;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.Set;
+import java.util.UUID;
+
+import static java.util.Objects.requireNonNull;
+
+public class RatisServer
+{
+    private final InternalNodeManager internalNodeManager;
+    private final String id;
+    private final String groupId;
+    private final int port;
+    private final String storageDir;
+
+    @Inject
+    public RatisServer(InternalNodeManager internalNodeManager, RaftConfig raftConfig)
+    {
+        this(
+                requireNonNull(internalNodeManager, "internalNodeManager is null"),
+                internalNodeManager.getCurrentNode().getNodeIdentifier(),
+                requireNonNull(raftConfig, "raftConfig is null").getGroupId(),
+                raftConfig.getPort(),
+                raftConfig.getStorageDir());
+    }
+
+    public RatisServer(InternalNodeManager internalNodeManager, String id, String groupId, int port, String storageDir)
+    {
+        this.internalNodeManager = internalNodeManager;
+        this.id = id;
+        this.groupId = groupId;
+        this.port = port;
+        this.storageDir = storageDir;
+    }
+
+    public RaftPeer[] getPeers()
+    {
+        Set<InternalNode> resourceManagers = internalNodeManager.getResourceManagers();
+        return resourceManagers.stream().map(resourceManager -> {
+            RaftPeer.Builder builder = RaftPeer.newBuilder();
+            builder.setId(RaftPeerId.valueOf(resourceManager.getNodeIdentifier()))
+                   .setAddress(resourceManager.getHost() + ":" + resourceManager.getRaftPort().getAsInt());
+            return builder.build();
+        }).toArray(RaftPeer[]::new);
+    }
+
+    @PostConstruct
+    public void start()
+            throws Exception
+    {
+        run();
+    }
+
+    public void run() throws Exception
+    {
+        RaftPeerId peerId = RaftPeerId.valueOf(id);
+        RaftProperties properties = new RaftProperties();
+
+        GrpcConfigKeys.Server.setPort(properties, port);
+
+        File storage = new File(storageDir + "/" + id);
+        RaftServerConfigKeys.setStorageDir(properties, Collections.singletonList(storage));
+
+        final RaftGroup raftGroup = RaftGroup.valueOf(RaftGroupId.valueOf(UUID.nameUUIDFromBytes(groupId.getBytes())), getPeers());
```

Comment:
Can you help me understand the logic around `UUID.nameUUIDFromBytes` .
Since groupId is something that is set from static config, how would it be a UUID?

Reply:
RaftGroupId's are of type UUID so even though the groupId is set as a string from the config it needs to be converted to a UUID in order to build the RaftGroup.

Reply:
Is the expectation that in the static config we set a UUID string and then convert it back to UUID object using `UUID.nameUUIDFromBytes` . If that is the case wondering why it should be mentioned in config. Can't we generate a group ID as a `UUID.randomUUID` in code. In other words, should the group ID be even part of the config.

Reply:
The groupId would also be needed when we build the RaftClient for the coordinator as the servers and client belong to the same RaftGroup.